File: hyuns/dfs/Programmers_네트워크.py
# 2차 풀이 - 실패
def dfs(x, computers, visit):
    for i in range(len(computers)):
        if visit[i] == False and computers[x][i] == 1:
            visit[i] = True
            dfs(i, computers, visit)

def solution(n, computers):
    answer = 0
    visit = [False for _ in range(n)]
    
    for i in range(n):
        if visit[i] == False:
            visit[i] = True
            dfs(i, computers, visit)
            answer += 1

    return answer

print(solution(4, [[1, 1, 0, 1], [1, 1, 0, 0], [0, 0, 1, 1], [1, 0, 1, 1]]))



# 1차 풀이
# def dfs(computers, index, visit):
#     for i in range(len(computers[index])):
#         if computers[index][i] == 1 and visit[i] == False:
#             visit[i] = True
#             dfs(computers, i, visit)
#     return None
# def solution(n, computers):
#     answer = 0
#     visit = [False for _ in range(n)]
    
#     for i in range(n):
#         if visit[i] == False:
#             dfs(computers, i, visit)
#             answer += 1
#     return answer

File: hyuns/practice_js/Programmers_네트워크.js
function dfs(i, visit, computers){
    for (var j=0; j<computers.length; j++){
        if (i===j){
            continue;
        }
        if (computers[i][j]==1 && visit[j] === false){
            visit[j] = true;
            dfs(j, visit, computers);
        }
    }
    
}
function solution(n, computers) {
    var answer = 0;
    let visit = []
    for (var i=1; i<=n; i++){
        visit.push(false)
    }

    
    for (var i=0; i<n; i++){
        if (visit[i] === false){
            visit[i] = true
            dfs(i, visit, computers)
            answer += 1
        }
    }
    return answer;
}

numbers = [1,1,1,1,1]
target = 3

console.log(solution(numbers, target))